AWS WAF module custom waiter

ryansb opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
WAF modules don't wait for their change to propagate fully, this change adds a custom waiter (dependent on #36922) to make sure the changes fully propagate before the module returns. 


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
aws_waf modules

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

ryansb committed:
# This is a combination of 8 commits.
# This is the 1st commit message:

Spike of custom route waiter

# This is the commit message #2:

Name waiter correctly

# This is the commit message #3:

Switch to new waiter, and refactor waiters to be specced per service

# This is the commit message #4:

PEP8

# This is the commit message #5:

Require boto3/botocore at runtime instead of at import

# This is the commit message #6:

PEP8

# This is the commit message #7:

PEP8

# This is the commit message #8:

Consolidate waiters to a single file

ryansb committed:
Add waiter message with token ID

ryansb committed:
WAF data model refactor

s-hertel commented:
This should fix the sporadic error here: https://github.com/ansible/ansible/pull/36603#issuecomment-368149668

ryansb committed:
Fix ref to self.client

ryansb committed:
Add custom waiters to aws_waf_web_acl

ryansb committed:
Allow add/remove rule tasks to operate in parallel, then wait for their change tokens to complete

willthames commented:
I rebased this on the latest devel now that #36922 has been merged

s-hertel commented:
I'm still triggering this occasional failure (with this patch), so I think the waiters may not be in the right place. Still trying to debug.
```
TASK [aws_waf_web_acl : check that a new pattern is created (because the first pattern should have been deleted once unused)] ***
task path: /Users/shertel/Workspace/ansible/test/integration/targets/aws_waf_web_acl/tasks/main.yml:229
fatal: [localhost]: FAILED! => {
    "assertion": "recreate_waf_regex_condition.condition.regex_match_tuples[0].regex_pattern_set_id != create_waf_regex_condition.condition.regex_match_tuples[0].regex_pattern_set_id\n",
    "changed": false,
    "evaluated_to": false
}
```

ryansb committed:
Move waiter into run_func_with_change_token_backoff since it is generic to all WAF update operations

ryansb committed:
Add waiter

Add waiter for WAF change tokens

Working waiter for waf_condition module

Add support for waiters to waf_rule

ryansb committed:
Wait for deletes on waf_web_acl

ryansb commented:
It appears these patches still don't fix the underlying issue. Continuing to tweak where waits happen to see if it's related to that, or if the integration test is making a faulty assumption somewhere.

ryansb committed:
NOMERGE:Try to *ALWAYS* wait for WAF to do things

ryansb committed:
Remove always-wait

ryansb committed:
Oops variables

ryansb commented:
@willthames can you take another look at this? I've used the custom waiter in all the final-step modifications in the WAF modules, so now they always wait for the changetoken to change to the Completed state. 

The downside is that it makes all the modules significantly slower, so the integration tests for WAF web ACL now take about an hour. The up side here is that now the modules are *always* leaving the WAF resources in a consistent state, so the test flakiness issue is resolved.

ryansb committed:
Delete extra debug statements in integration tests

willthames commented:
If this greatly changes the length of time taken to use the modules, should we add a `wait` parameter to the modules?

willthames commented:
Rebased on latest waiter changes, will rerun full test suite now before merge.

willthames commented on lib/ansible/modules/cloud/amazon/aws_waf_condition.py:
```diff
@@ -580,7 +583,13 @@ def find_and_update_condition(self, condition_set_id):
+            except botocore.exceptions.WaiterError as e:
+                self.module.fail_json_aws(
+                    e,
+                    msg='Change token did not sync, or took too long %s : %s' % (
```
This line is problematic because `result` does not get set if the exception occurs.

Is there a reason we catch the WaiterError in aws_waf_condition but not in the change to the other two modules?

ryansb replied:
Oh, good catch. I think that's why I took it out in the other two. I'll patch this.

ryansb committed:
Remove waiter retry catch
